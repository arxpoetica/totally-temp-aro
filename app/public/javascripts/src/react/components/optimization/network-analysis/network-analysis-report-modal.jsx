import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import reduxStore from '../../../../redux-store'
import wrapComponentWithProvider from '../../../common/provider-wrapped-component'
import NetworkAnalysisActions from './network-analysis-actions'

// An internal class for rendering a single row in the reports table
class NetworkAnalysisReportRow extends Component {
  constructor (props) {
    super(props)
    this.state = {
      selectedFormat: this.props.mediaTypes[0]
    }
  }

  render () {
    // "(new Date()).toISOString().split('T')[0]" will give "YYYY-MM-DD"
    const downloadFileName = `${(new Date()).toISOString().split('T')[0]}_${this.props.reportName}.${this.state.selectedFormat}`
    return <tr>
      <td>{this.props.displayName}</td>
      <td>
        <select
          className='form-control form-control-sm'
          value={this.state.selectedFormat}
          onChange={event => this.setState({ selectedFormat: event.target.value })}>
          { this.props.mediaTypes.map(item => <option key={item} value={item}>{item}</option>)}
        </select>
      </td>
      <td style={{ textAlign: 'center' }}>
        <a className='btn btn-sm btn-light'
          href={`/service-download-file/${downloadFileName}/v2/report-extended/${this.props.reportId}/${this.props.planId}.${this.state.selectedFormat}`}
          download>
          <i className='fa fa-download' /> Download
        </a>
      </td>
    </tr>
  }
}
NetworkAnalysisReportRow.propTypes = {
  planId: PropTypes.number,
  reportId: PropTypes.number,
  reportName: PropTypes.string,
  displayName: PropTypes.string,
  mediaTypes: PropTypes.array
}

export class NetworkAnalysisReportModal extends Component {
  render () {
    return <div
      id='modalNetworkAnalysisReports'
      className={`modal fade ${this.props.showReportModal ? 'show' : ''}`}
      role='dialog'
      style={{ display: this.props.showReportModal ? 'block' : 'none' }}
    >
      <div className='modal-dialog'>
        <div className='modal-content'>
          <div className='modal-header'>
            <h4 className='modal-title'>Network Analysis Reports</h4>
            <button
              type='button'
              onClick={() => this.props.showOrHideReportModal(false)}
              className='close'
            >
              &times;
            </button>
          </div>
          <div className='modal-body'>
            <table className='table table-sm table-striped'>
              <thead>
                <tr>
                  <th>Report</th>
                  <th>Format</th>
                  <th />
                </tr>
              </thead>
              <tbody>
                {
                  (this.props.reportsMetaData || []).map((report, index) => (
                    <NetworkAnalysisReportRow
                      key={index}
                      reportId={report.id}
                      reportName={report.name}
                      displayName={report.displayName}
                      mediaTypes={report.media_types}
                      planId={this.props.planId} />
                  ))
                }
              </tbody>
            </table>
          </div>
          <div className='modal-footer'>
            <button
              onClick={() => this.props.showOrHideReportModal(false)}
              type='button'
              className='btn btn-light'
            >
              Close
            </button>
          </div>
        </div>
      </div>
    </div>
  }

  componentWillReceiveProps (nextProps) {
    // If the modal dialog is going from hidden to visible state, load reports metadata
    if (!this.props.showReportModal && nextProps.showReportModal) {
      this.props.loadReportsMetaData()
    }
  }
}

NetworkAnalysisReportModal.propTypes = {
  planId: PropTypes.number,
  reportsMetaData: PropTypes.array,
  showReportModal: PropTypes.bool
}

const mapStateToProps = state => ({
  planId: state.plan.activePlan.id,
  reportsMetaData: state.optimization.networkAnalysis.reportsMetaData,
  showReportModal: state.optimization.networkAnalysis.showReportModal
})

const mapDispatchToProps = dispatch => ({
  loadReportsMetaData: () => dispatch(NetworkAnalysisActions.loadReportsMetaData()),
  showOrHideReportModal: showReportModal => dispatch(NetworkAnalysisActions.showOrHideReportModal(showReportModal))
})

const NetworkAnalysisReportModalComponent = wrapComponentWithProvider(reduxStore, NetworkAnalysisReportModal, mapStateToProps, mapDispatchToProps)
export default NetworkAnalysisReportModalComponent
