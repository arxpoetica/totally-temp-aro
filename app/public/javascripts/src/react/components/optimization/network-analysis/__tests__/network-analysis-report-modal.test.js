/* global test expect jest */
import React from 'react'
import { shallow } from 'enzyme'
import { NetworkAnalysisReportModal, NetworkAnalysisReportRow } from '../network-analysis-report-modal'

// -----------------------------------------------------------------------------
test('When modal is hidden', () => {
  const component = shallow(
    <NetworkAnalysisReportModal
      planId={100}
      showReportModal={false}
    />
  )
  expect(component).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When modal is shown', () => {
  const component = shallow(
    <NetworkAnalysisReportModal
      planId={100}
      showReportModal
    />
  )
  expect(component).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When modal is initially hidden and then shown', () => {
  const mockLoadReportsMetaData = jest.fn()
  const component = shallow(
    <NetworkAnalysisReportModal
      planId={100}
      showReportModal={false}
      loadReportsMetaData={mockLoadReportsMetaData}
    />
  )
  expect(component).toMatchSnapshot()
  component.setProps({ showReportModal: true })
  expect(mockLoadReportsMetaData).toHaveBeenCalled()
  expect(component).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When reports metadata is null', () => {
  const component = shallow(
    <NetworkAnalysisReportModal
      planId={100}
      reportsMetaData={null}
    />
  )
  expect(component).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('When reports metadata is defined', () => {
  const reportsMetaData = [
    {
      displayName: 'Optimization Analysis',
      id: 4,
      media_types: ['csv', 'json', 'xls'],
      name: 'optimization_analysis',
      reportType: 'NETWORK_ANALYSIS'
    }
  ]
  const component = shallow(
    <NetworkAnalysisReportModal
      planId={100}
      reportsMetaData={reportsMetaData}
    />
  )
  expect(component).toMatchSnapshot()
})

// -----------------------------------------------------------------------------
test('NetworkAnalysisReportRow render', () => {
  // Mock the Date.now() function so that the test is determinstic
  Date.now = jest.fn(() => 1557383970443)
  const component = shallow(
    <NetworkAnalysisReportRow
      planId={100}
      reportId={287}
      reportName={'optimization_analysis_report'}
      displayName={'Optimization analysis report'}
      mediaTypes={['csv', 'json', 'xls']}
    />
  )
  expect(component).toMatchSnapshot()
})
