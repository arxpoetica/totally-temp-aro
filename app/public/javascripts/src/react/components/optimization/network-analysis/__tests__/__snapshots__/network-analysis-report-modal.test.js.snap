// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkAnalysisReportRow render 1`] = `
<tr>
  <td>
    Optimization analysis report
  </td>
  <td>
    <select
      className="form-control form-control-sm"
      onChange={[Function]}
      value="csv"
    >
      <option
        key="csv"
        value="csv"
      >
        csv
      </option>
      <option
        key="json"
        value="json"
      >
        json
      </option>
      <option
        key="xls"
        value="xls"
      >
        xls
      </option>
    </select>
  </td>
  <td
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <a
      className="btn btn-sm btn-light"
      download={true}
      href="/service-download-file/2019-05-08_optimization_analysis_report.csv/v2/report-extended/287/100.csv"
    >
      <i
        className="fa fa-download"
      />
       Download
    </a>
  </td>
</tr>
`;

exports[`When modal is hidden 1`] = `
<div
  className="modal fade "
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "none",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`When modal is initially hidden and then shown 1`] = `
<div
  className="modal fade "
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "none",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`When modal is initially hidden and then shown 2`] = `
<div
  className="modal fade show"
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`When modal is shown 1`] = `
<div
  className="modal fade show"
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "block",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`When reports metadata is defined 1`] = `
<div
  className="modal fade "
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "none",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <NetworkAnalysisReportRow
              displayName="Optimization Analysis"
              key="0"
              mediaTypes={
                Array [
                  "csv",
                  "json",
                  "xls",
                ]
              }
              planId={100}
              reportId={4}
              reportName="optimization_analysis"
            />
          </tbody>
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`When reports metadata is null 1`] = `
<div
  className="modal fade "
  id="modalNetworkAnalysisReports"
  role="dialog"
  style={
    Object {
      "display": "none",
    }
  }
>
  <div
    className="modal-dialog"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <h4
          className="modal-title"
        >
          Network Analysis Reports
        </h4>
        <button
          className="close"
          id="btnHeaderHideReportModal"
          onClick={[Function]}
          type="button"
        >
          ×
        </button>
      </div>
      <div
        className="modal-body"
      >
        <table
          className="table table-sm table-striped"
        >
          <thead>
            <tr>
              <th>
                Report
              </th>
              <th>
                Format
              </th>
              <th />
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
      <div
        className="modal-footer"
      >
        <button
          className="btn btn-light"
          id="btnHideReportModal"
          onClick={[Function]}
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
